### PR TITLE
XFEM geometric cut switching material system

### LINKED ISSUE #17041: Refactoring the levelsetBimaterial System
## Reason

Bimaterial isn't enough any more, we need to switch among multiple materials.

## Design

Generalize the level-set idea to all geometric cut userobjects. Move the mapping logic into a ComboCutUserobject.

## Impact

Some minor changes to internal XFEM APIs.
New classes and new input file syntax.

### PR DESCRIPTION
**Pull request by hugary1995**
Add SwitchingMaterial, ComboCutUserObject, GeometricCutSubdomainIDAux and tests

close #17041

### CODE REVIEW

**Comment by hugary1995** [comment]
@bwspenc and @jiangwen84 this is part 2 of the refactoring.

[https://github.com/idaholab/moose/blob/2bee01da8fb7d88bb9e77e98d49cdaf8dbd601eb/modules/xfem/test/tests/switching_material/two_cuts_moving.i](https://github.com/idaholab/moose/blob/2bee01da8fb7d88bb9e77e98d49cdaf8dbd601eb/modules/xfem/test/tests/switching_material/two_cuts_moving.i) pretty much demonstrates all components of the new design:
```
[Problem]
  solve = false
[]

[XFEM]
  qrule = volfrac
  output_cut_plane = true
[]

[UserObjects]
  [cut1]
    type = LevelSetCutUserObject
    level_set_var = phi1
    negative_id = 1
    positive_id = 33
    execute_on = NONE
  []
  [cut2]
    type = LevelSetCutUserObject
    level_set_var = phi2
    negative_id = 5
    positive_id = 16
    execute_on = NONE
  []
  [combo]
    type = ComboCutUserObject
    cuts = 'cut1 cut2'
    keys = '1 5
            1 16
            33 5
            33 16'
    vals = '1 3 5 7'
    heal_always = true
  []
[]

[Mesh]
  [square]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 5
    ny = 5
  []
[]

[AuxVariables]
  [phi1]
  []
  [phi2]
  []
  [cut1_id]
    order = CONSTANT
    family = MONOMIAL
  []
  [cut2_id]
    order = CONSTANT
    family = MONOMIAL
  []
  [combo_id]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[AuxKernels]
  [phi1]
    type = FunctionAux
    variable = phi1
    function = 'x-0.213-t'
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
  [phi2]
    type = FunctionAux
    variable = phi2
    function = 'x-0.728+t'
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
  [cut1_id]
    type = GeometricCutSubdomainIDAux
    variable = cut1_id
    cut = cut1
  []
  [cut2_id]
    type = GeometricCutSubdomainIDAux
    variable = cut2_id
    cut = cut2
  []
  [combo_id]
    type = GeometricCutSubdomainIDAux
    variable = combo_id
    cut = combo
  []
[]

[Materials]
  [diffusivity_A]
    type = GenericConstantMaterial
    prop_names = 'A_D'
    prop_values = '5'
  []
  [diffusivity_B]
    type = GenericConstantMaterial
    prop_names = 'B_D'
    prop_values = '0.5'
  []
  [diffusivity_C]
    type = GenericConstantMaterial
    prop_names = 'C_D'
    prop_values = '12'
  []
  [diffusivity_D]
    type = GenericConstantMaterial
    prop_names = 'D_D'
    prop_values = '9'
  []
  [diff_combined]
    type = GeometricCutSwitchingMaterialReal
    base_name_keys = '1 3 5 7'
    base_name_vals = 'A B C D'
    prop_name = D
    geometric_cut_userobject = combo
    outputs = 'exodus'
    output_properties = 'D'
  []
[]

[Executioner]
  type = Transient

  dt = 0.25
  num_steps = 2

  max_xfem_update = 1
[]

[Outputs]
  exodus = true
[]
```
We can discuss how to fit the 3D cut (using a 2D mesh) into this new system.

File: modules/xfem/doc/content/source/userobjects/ComboCutUserObject.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# ComboCutUserObject
+
+!syntax description /UserObjects/ComboCutUserObject
+
+## Overview
+
+The `ComboCutUserObject` combines multiple `GeometricCutUserObject`s to create a composite cut. Currently, all `GeometricCutUserObject`s are assumed to be non-overlapping, i.e. each element can only be cut by one geometric cut.
```
Comment (bwspenc) [inline modules/xfem/doc/content/source/userobjects/ComboCutUserObject.md:7]:
```
The `ComboCutUserObject` combines multiple instances of `GeometricCutUserObject` to define the geometric cut subdomain ID for all portions of the domain based on which side that point is located in relative to multiple cuts defined by `GeometricCutUserObject` instances. Currently, all `GeometricCutUserObject`s are assumed to be non-overlapping, i.e. each element can only be cut by one geometric cut.

File: modules/xfem/doc/content/source/userobjects/ComboCutUserObject.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+Each `GeometricCutUserObject` may have defined its own set of `GeometricCutSubdomainID`s, and the `ComboCutUserObject` follows the key-value dictionary provided by the user to assign new `GeometricCutSubdomainID` to each of the cut-subdomain.
```
Comment (bwspenc) [inline modules/xfem/doc/content/source/userobjects/ComboCutUserObject.md:9]: to assign new -> to assign a

Comment (bwspenc) [reply modules/xfem/doc/content/source/userobjects/ComboCutUserObject.md:9]: cut-subdomain -> geometric cut subdomains.

File: modules/xfem/include/base/XFEM.h [hunk 3]
```diff
@@ -323,12 +334,17 @@ class XFEM : public XFEMInterface
...
    */
-  std::map<const Elem *, std::map<const GeometricCutSubdomainID, const Elem *>> _healed_elems;
-  std::map<const Elem *, std::map<const GeometricCutSubdomainID, bool>>
-      _healed_material_properties_used;
-
-  /// healed geometric cuts
-  std::map<const Elem *, const GeometricCutUserObject *> _healed_cuts;
+  typedef std::tuple<const Elem *, const GeometricCutUserObject *, GeometricCutSubdomainID>
+      GeometricCutElemInfo;
+  std::map<const Elem *, GeometricCutElemInfo> _geom_cut_elems;
```
Comment (bwspenc) [inline modules/xfem/include/base/XFEM.h:339]: Please add doxygen one-liners for each of these

File: modules/xfem/include/base/XFEM.h [hunk 4]
```diff
@@ -424,26 +440,19 @@ class XFEM : public XFEMInterface
...
-                                       const Elem * cut_elem,
-                                       const GeometricCutUserObject * gcuo);
+  void setMaterialPropertiesForElement(const Elem * elem, const Elem * elem_from);
 
   /**
-   * Determine which cut subdomain of the element belongs to relative to the cut
-   * @param parent_elem The parent element
-   * @param cut_elem    The element being cut
-   * @param gcuo        The GeometricCutUserObject for the cut
+   * Randomly pick one node on the element that is in the physical domain
```
Comment (bwspenc) [inline modules/xfem/include/base/XFEM.h:453]: I'd reword this as:

Return the first node in the provided element that is found to be in the physical domain

...because it's not actually random.

File: modules/xfem/include/materials/GeometricCutSwitchingMaterial.h [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+#include "Material.h"
+#include "XFEM.h"
+#include "ADRankTwoTensorForward.h"
+#include "ADRankThreeTensorForward.h"
+#include "ADRankFourTensorForward.h"
+
+/**
+ * switches between materials in a multi-material system where the
+ * interfaces are defined by multiple level set functions.
```
Comment (bwspenc) [inline modules/xfem/include/materials/GeometricCutSwitchingMaterial.h:20]: multiple geometric cut objects, right?

File: modules/xfem/include/materials/GeometricCutSwitchingMaterial.h [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+template <typename T, bool is_ad>
+class GeometricCutSwitchingMaterialTempl : public Material
+{
+public:
+  static InputParameters validParams();
+
+  GeometricCutSwitchingMaterialTempl(const InputParameters & parameters);
+
+protected:
+  // At the time of initiating this switching material, the initeQpStatefulProperties methods of the
```
Comment (bwspenc) [inline modules/xfem/include/materials/GeometricCutSwitchingMaterial.h:31]: initiating -> initializing

**Review by bwspenc — CHANGES_REQUESTED** [review]
I haven't had a chance to go through all of this yet, and have to sign off for the day, but here are some relatively minor things to address. I think this is really close to being ready to go.

**Comment by hugary1995** [comment]
@bwspenc I fixed all that.

**Comment by hugary1995** [comment]
@jiangwen84 @dschwen Could you please review this PR when you get a chance? There are some significant changes/additions to the XFEM module, so I think it's better to have more eyes on it.

The algorithmic flow is summarized here [https://mooseframework.inl.gov/docs/PRs/17042/site/modules/xfem/theory/embedded_interface.html](https://mooseframework.inl.gov/docs/PRs/17042/site/modules/xfem/theory/embedded_interface.html), and I tried my best to make the source code adequately commented.

File: modules/xfem/include/base/XFEM.h [hunk 7]
```diff
@@ -417,33 +474,59 @@ class XFEM : public XFEMInterface
...
-   * @param gcuo        The GeometricCutUserObject for the cut
+   * Return the first node in the provided element that is found to be in the physical domain
+   * @param e  Constant pointer to the child element
+   * @param e0 Constant pointer to the parent element whose nodes will be querried
+   * @return A constant pointer to the node
    */
-  GeometricCutSubdomainID getGeometricCutSubdomainID(const Elem * parent_elem,
-                                                     const Elem * cut_elem,
-                                                     const GeometricCutUserObject * gcuo) const;
+  const Node * pickOnePhysicalNode(const Elem * e, const Elem * e0) const;
```
Comment (bwspenc) [inline modules/xfem/include/base/XFEM.h:531]: It might be clearer to rename this to `pickFirstPhysicalNode`, but do as you see fit.

File: modules/xfem/src/materials/GeometricCutSwitchingMaterial.C [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+template <typename T, bool is_ad>
+void
+GeometricCutSwitchingMaterialTempl<T, is_ad>::computeProperties()
+{
+  GeometricCutSubdomainID key = _xfem->getGeometricCutSubdomainID(_cut, _current_elem);
+
+  // We may run into situations where the key doesn't exist in the base_name_map. This may happen
+  // when the problem is not well defined, or the level-set variables are very nonlinear, so that
+  // the combination of levelset signs are not provided. Cutting the timestep may help, so we throw
```
Comment (bwspenc) [inline modules/xfem/src/materials/GeometricCutSwitchingMaterial.C:79]: Would it make sense to query the geometric cut userobject for the list of keys it can provide at initialization and make sure we cover all of the possibilities then rather than during the computation?

File: modules/xfem/src/materials/GeometricCutSwitchingMaterial.C [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+template <typename T, bool is_ad>
+InputParameters
+GeometricCutSwitchingMaterialTempl<T, is_ad>::validParams()
+{
+  InputParameters params = Material::validParams();
+  params.addClassDescription("Switch the material property based on the GeometricCutSubdomainID.");
+  params.addRequiredParam<UserObjectName>("geometric_cut_userobject",
+                                          "The geometric cut userobject");
+  params.addRequiredParam<std::vector<GeometricCutSubdomainID>>(
+      "base_name_keys",
```
Comment (bwspenc) [inline modules/xfem/src/materials/GeometricCutSwitchingMaterial.C:31]: I think it might be clearer to call these parameters `geometric_cut_subdomain_ids` and `base_names`. We can discuss, though, if you don't like those suggestions.

File: modules/xfem/src/userobjects/ComboCutUserObject.C [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+InputParameters
+ComboCutUserObject::validParams()
+{
+  InputParameters params = GeometricCutUserObject::validParams();
+  params.addClassDescription("Combine multiple geometric cut userobjects.");
+  params.addRequiredParam<std::vector<UserObjectName>>("cuts",
+                                                       "Vector of cut userobjects to combine");
+  params.addRequiredParam<std::vector<GeometricCutSubdomainID>>(
+      "keys", "Possible combinations of the geometric cut subdomain IDs");
```
Comment (bwspenc) [inline modules/xfem/src/userobjects/ComboCutUserObject.C:22]: Sorry, I'm not crazy about `keys` and `vals` either. Maybe `input_cut_ids` and `output_cut_ids`?

File: modules/xfem/src/userobjects/ComboCutUserObject.C [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    _cuts[i] = &getUserObjectByName<GeometricCutUserObject>(_cut_names[i]);
+
+  buildMap();
+}
+
+void
+ComboCutUserObject::buildMap()
+{
+  if (_keys.size() % _num_cuts != 0)
+    mooseError("Expect multiples of ", _num_cuts, " keys, but got ", _keys.size(), " keys.");
```
Comment (bwspenc) [inline modules/xfem/src/userobjects/ComboCutUserObject.C:47]: 'Expect' -> 'Expected'

File: modules/xfem/src/userobjects/ComboCutUserObject.C [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+  unsigned int num_combos = _keys.size() / _num_cuts;
+
+  if (_vals.size() != num_combos)
+    mooseError("Expect one value for each of the combination, got ",
```
Comment (bwspenc) [inline modules/xfem/src/userobjects/ComboCutUserObject.C:52]: 'Expect'->'Expected'

File: modules/xfem/src/userobjects/ComboCutUserObject.C [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+  return false;
+}
+
+GeometricCutSubdomainID
+ComboCutUserObject::getGeometricCutSubdomainID(const Node * node) const
+{
+  std::vector<GeometricCutSubdomainID> combo_key;
+  for (auto cut : _cuts)
+    combo_key.push_back(cut->getGeometricCutSubdomainID(node));
+  return _combo_ids.at(combo_key);
```
Comment (bwspenc) [inline modules/xfem/src/userobjects/ComboCutUserObject.C:118]: It seems like there's a chance that a particular combination might not exist, so we might need to handle that error. I'm saying that because I don't think the check on line 46 of this class exhaustively checks that you have every combination covered.

**Review by bwspenc — CHANGES_REQUESTED** [review]
Overall, looks really good!  I'm excited to get this in. Just some really minor picky stuff.

File: modules/xfem/src/materials/GeometricCutSwitchingMaterial.C (hunk 8 above)
Comment (hugary1995) [reply modules/xfem/src/materials/GeometricCutSwitchingMaterial.C:79]: Yeah that's the right thing to do. This check was carried over from my old design.

File: modules/xfem/src/materials/GeometricCutSwitchingMaterial.C (hunk 9 above)
Comment (hugary1995) [reply modules/xfem/src/materials/GeometricCutSwitchingMaterial.C:31]: `geometric_cut_subdomain_ids` and `base_names` are better. Again, these parameter names were carried over from the previous design where the key/value naming made more sense.

File: modules/xfem/src/userobjects/ComboCutUserObject.C (hunk 13 above)
Comment (hugary1995) [reply modules/xfem/src/userobjects/ComboCutUserObject.C:118]: Yeah, that check in `GeometricCutSwitchingMaterial.C` should be moved here.

File: modules/xfem/src/userobjects/ComboCutUserObject.C (hunk 10 above)
Comment (hugary1995) [reply modules/xfem/src/userobjects/ComboCutUserObject.C:22]: I want to somehow convey the message that we are doing some sort of look-up here. What are synonyms for "key-value"?

File: modules/xfem/include/base/XFEM.h [hunk 14]
```diff
@@ -40,6 +40,52 @@ enum XFEM_QRULE
...
   DIRECT
 };
+
+/**
+ * Convenient typedef for local storage of stateful material properties. The first (component 0)
+ * entry in the CachedMaterialProperties is a map for old material properties. The second
+ * (component 1) entry is a map for older material properties. The second entry will be empty if
+ * the material storage doesn't have older material properties.
+ */
+typedef std::array<std::unordered_map<unsigned int, std::string>, 2> CachedMaterialProperties;
```
[the remark this answers is not in the document]
Comment (hugary1995) [reply modules/xfem/include/base/XFEM.h:50]: If there is not stateful properties, the `_geom_cut_elems` storage will be empty. See `XFEM::storeMaterialPropertiesForElementHelper`.

File: modules/xfem/include/userobjects/ComboCutUserObject.h [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+class ComboCutUserObject : public GeometricCutUserObject
+{
+public:
+  static InputParameters validParams();
+
+  ComboCutUserObject(const InputParameters & parameters);
+
+  /**
+   * Loop over all the provided GeometricCutUserObjects, fill the data structures based on the first
+   * cut that wants to cut this 2D element.
```
[the remark this answers is not in the document]
Comment (hugary1995) [reply modules/xfem/include/userobjects/ComboCutUserObject.h:23]: Unfortunately it is (as of now). If multiple cuts want to cut the same element, how should we handle it?

File: modules/xfem/src/materials/GeometricCutSwitchingMaterial.C (hunk 8 above)
Comment (hugary1995) [reply modules/xfem/src/materials/GeometricCutSwitchingMaterial.C]: On a second thought, it's actually difficult to do the check at construction time because I can't query all the possible cut subdomain IDs for a given geometric cut userobject. Unless we require all geometric cut userobjects to explicit define their possible cut subdomain IDs.

File: modules/xfem/doc/content/modules/xfem/theory/embedded_interface.md [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+XFEM is used to insert strong discontinuities along each interface. If desired, solution continuity across the interfaces can be enforced using either a penalty constraint or the Nitsche's method.  
+
+## Moving interface
+
+To represent the movement of the interface, the user only need to update the interface location defined by the corresponding `GeometricCutUserObject`. The XFEM module handles the update of the interface location using the so-called *healing-and-re-cut* algorithm. In a nutshell, every time a interface location is updated, its corresponding cut (as well as the underlying strong discontinuity) is completed removed (thereby termed with "healing"), and the updated interface is marked and re-cut following the [mesh cutting algorithm](xfem/theory/theory.md#mesh_cutting_alg) using the updated interface location.
```
Comment (bwspenc) [inline modules/xfem/doc/content/modules/xfem/theory/embedded_interface.md:11]: the user only need -> the user only needs

File: modules/xfem/doc/content/modules/xfem/theory/embedded_interface.md [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+
+The meshes and the `CEI` data structures are shown in [healing_recut_1], [healing_recut_2] and [healing_recut_3], for the first, the second, and the third time step, respectively.
+
+!media media/xfem/healing_recut_1.png
+       id=healing_recut_1
+       style=width:100%;padding:20px;
+       caption=The *healing-and-re-cut* algorithm at time step 1. For demonstration purposes, the `CEI`s are denoted using element IDs and the cut ID, while in the actual implementation pointers are used to store these information.
+
+During the first solve in the first time step, the mesh is not yet cut by XFEM, and the `CEI`s are empty. Therefore, nothing happens during function call `healMesh`. The cut marks element 1 for cutting, and it is cut into children elements 3 and 4 during the function call `cutMeshWithEFA`. When child element 3 is created, its parent element (1), the correponding cut (1), and its cut subdomain (1) are stored into the current `CEI`, denoted here as $\{3,1,1,1\}$. Similarly, a `CEI` entry $\{4,1,1,2\}$ is stored for child element 4. The system is solved again after the interface has been introduced.
```
Comment (bwspenc) [inline modules/xfem/doc/content/modules/xfem/theory/embedded_interface.md:32]: healMesh -> XFEM::healMesh()

File: modules/xfem/doc/content/modules/xfem/theory/embedded_interface.md [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+!media media/xfem/healing_recut_2.png
+       id=healing_recut_2
+       style=width:100%;padding:20px;
+       caption=The *healing-and-re-cut* algorithm at time step 2. For demonstration purposes, the `CEI`s are denoted using element IDs and the cut ID, while in the actual implementation pointers are used to store these information.
+
+In the second time step, the system is first solved with updated boundary conditions, during which the signed distance function may be updated. In this case, the interface moves towards the right. During the function call `healMesh`, the cut is removed by deleting element 4. Since child element 3 is now treated as a potential parent element, the corresponding entries in `CEI` are updated to reflect that.
```
Comment (bwspenc) [inline modules/xfem/doc/content/modules/xfem/theory/embedded_interface.md:39]: healMesh->XFEM::healMesh()

File: modules/xfem/include/userobjects/ComboCutUserObject.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (bwspenc) [reply modules/xfem/include/userobjects/ComboCutUserObject.h:23]: If we get a reasonable error in that case, that's fine for now. I assume that the cutter objects are doing the right thing. We just need to make sure we don't forget this when we add the ability for multiple cuts.

File: modules/xfem/test/tests/switching_material/two_cuts_moving.i [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+    type = LevelSetCutUserObject
+    level_set_var = phi2
+    negative_id = 5
+    positive_id = 16
+    execute_on = NONE
+  []
+  [combo]
+    type = ComboCutUserObject
+    geometric_cut_userobjects = 'cut1 cut2'
+    cut_subdomain_combinations = '1 5 1 16 33 5 33 16'
```
Comment (bwspenc) [inline modules/xfem/test/tests/switching_material/two_cuts_moving.i:28]: We should probably break all of these up into separate lines for clarity (in this and the other input files). I'm guessing that hit format doesn't play nicely with that, though.

**Review by bwspenc — CHANGES_REQUESTED** [review]
Thanks for making those changes! I like the new syntax. Just a couple of really minor things that I saw in a second pass through. I'm not sure whether we're good to merge things just yet, but fix those up, and we'll have this all ready to go early next week when things are moving again.

**Comment by hugary1995** [comment]
Thanks Ben, I fixed all that. I also removed all the `Real time` arguments in all `GeometricCutUserObject`s -- it inherits `TransientInterface` so that seems redundant to me.

File: modules/xfem/test/tests/switching_material/two_cuts_moving.i (hunk 19 above)
Comment (hugary1995) [reply modules/xfem/test/tests/switching_material/two_cuts_moving.i]: This is now taking vector of vector of IDs per your suggestion. Input files are also updated.

**Review by bwspenc — APPROVED** [review]
Looks good, thanks!
